src/xmppjs/XJSGateway.ts

```diff
@@ -31,18 +39,24 @@ export interface RemoteGhostExtraData {
  * and XMPP.
  */
 export class XmppJsGateway implements IGateway {
-    // For storing room history, should be clipped at MAX_HISTORY per room.
-    private roomHistory: Map<string, [Element]>;
+    // For storing room history
+    private roomHistory: HistoryManager;
     // For storing requests to be responded to, like joins
     private stanzaCache: Map<string, Element>; // id -> stanza
     private presenceCache: PresenceCache;
     // Storing every XMPP user and their anonymous.
     private members: GatewayMUCMembership;
     constructor(private xmpp: XmppJsInstance, private registration: AutoRegistration, private config: IConfigBridge) {
-        this.roomHistory = new Map();
+        this.roomHistory = new HistoryManager(new MemoryStorage(50));
```

Comment:
Eventually, this should be configurable -- both the type of storage, and the parameters (e.g. size of storage).  I don't know if we should make it configurable now, or merge it as-is and make it configurable later.

Reply:
I'd suggest making a higher storage limit, e.g. 500 messages, and have a smaller default limit for sending history, e.g. 50 or 20.

That way, sane clients can fetch as much history as they need to fill the gap, whereas a reasonably short history is presented to newcomers.

Reply:
I went with 50 because I don't want to increase memory usage too much, and I wanted to start off cautiously.  It's something that we can tweak as we see how it goes.